### PR TITLE
[WIP] Missing parenthesis for object expressions

### PR DESCRIPTION
**Pull request by yamafaktory**
Will partially solve #243. Tests are still missing...

### CODE REVIEW

**Comment by vjeux** [comment]
I don't think that it is the right fix.

```js
({}.a());
({}).a();
```

outputs

```js
({}).a();
({}).a();
```

which is correct.

I think the issue is that needsParens function is not being called when `printMemberChain` prints things. cc @jlongster who has ideas about this.

**Comment by yamafaktory** [comment]
@vjeux oops, you're right it was only solving one of my test...

**Comment by jlongster** [comment]
If you look in the `genericPrint` print function, that's where it checks if an expression needs parentheses. This is the `print` function that's referenced throughout the printer, so basically every expression is checked to see if it needs parens and they are automatically added. This happens whenever you `print` part of the AST.

The problem is in places like `printMemberChain` when we split apart and AST node and print it ourselves rather than going through the normal `print` workflow. The parens checking is then skipped, so we probably just need to do it manually somewhere. I can't say any more without looking at it and I'm implementing something else right now, but I'll take a closer look soon.

**Comment by yamafaktory** [comment]
Ok, I think I get it, but I'm not really sure what is the best way to implement it. So I guess, I can check for extra -> parenthesized in each Node of type `ObjectExpression` (like in https://github.com/jlongster/prettier/blob/master/src/printer.js#L1965 but out of this loop of course) and then somehow need to print back the parens if detected.

**Comment by jlongster** [comment]
I'm not entirely sure what the best solution is yet either. I need to also solve it for my work that I'm doing for binary expression right now, because anywhere we manually traverse and print nodes we need this. So it might be best wait and see if I can solve both in the same pass.

**Comment by yamafaktory** [comment]
I don't get why running the tests via `npm test` and running a local test with `node ./bin/prettier.js ../test.js` (with the exact same example) gives me different ASTs (then of course no parens are added)... Locally the same test is at least partially working (`({ foo: "bar" }).a().b();`).

**Comment by vjeux** [comment]
`node ./bin/prettier.js ../test.js` uses babylon parser by default and `npm test` uses flow parser by default.

You can write `node ./bin/prettier.js ../test.js --flow-parser` to get flow on the command line and `run_spec(__dirname, {useFlowParser: false})` on npm to get babylon.

This is super weird and we should settle on one!

**Comment by yamafaktory** [comment]
@vjeux wow, thanks a lot for the guidance 👍. So actually the default one (used as bin) is the node one, but for the tests, the flow one is used. My question is: shouldn't the tests check both 😕?

**Comment by jlongster** [comment]
@yamafaktory Yep they should, it's on our todo list. The problem is babylon doesn't properly parse all the tests. But I don't think it should take too long to fix that.

File: src/fast-path.js [hunk 1]
```diff
@@ -374,6 +374,7 @@ FPp.needsParens = function(assumeExpressionContext) {
 
       switch (parent.type) {
         case "MemberExpression":
+        case "TaggedTemplateExpression":
```
Comment (jlongster) [inline src/fast-path.js]: This looks like it shouldn't be in this PR.

**Comment by jlongster** [comment]
Closing in lieu of #281 (see note there)

**Pull request closed without merging**